src/main/java/ca/uhn/fhir/jpa/starter/MyPlainProvider.java

```diff
@@ -0,0 +1,157 @@
+package ca.uhn.fhir.jpa.starter;
+
+import ca.uhn.fhir.context.FhirContext;
+import ca.uhn.fhir.parser.IParser;
+import ca.uhn.fhir.rest.annotation.Operation;
+import ca.uhn.fhir.rest.annotation.OptionalParam;
+import ca.uhn.fhir.rest.annotation.ResourceParam;
+import ca.uhn.fhir.rest.client.api.IGenericClient;
+import org.apache.http.client.ClientProtocolException;
+import org.apache.http.client.HttpClient;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.impl.client.HttpClients;
+import org.hl7.fhir.instance.model.api.IBaseResource;
+import org.hl7.fhir.r4.model.Bundle;
+import org.hl7.fhir.r4.model.MessageHeader;
+import org.hl7.fhir.r4.model.OperationOutcome;
+import org.hl7.fhir.r4.model.Reference;
+import org.hl7.fhir.r4.model.Resource;
+import org.hl7.fhir.r4.model.ResourceType;
+
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.util.List;
+
+public class MyPlainProvider {
+  @Operation(name="$process-message")
+  public IBaseResource processMessage(
+      @OptionalParam(name="async") String async,
+      @OptionalParam(name="response-url") String responseUrl,
+      @ResourceParam Bundle bundleR4
+  ) {
+    // validate the bundle
+    if (bundleR4.getType() != Bundle.BundleType.MESSAGE) {
+      OperationOutcome oo = new OperationOutcome();
+      oo.addIssue()
+          .setSeverity(OperationOutcome.IssueSeverity.ERROR)
+          .setDiagnostics("Bundle is not of type 'message'");
+      return oo;
+    }
+
+    if (!bundleR4.hasEntry()) {
+      OperationOutcome oo = new OperationOutcome();
+      oo.addIssue()
+          .setSeverity(OperationOutcome.IssueSeverity.ERROR)
+          .setDiagnostics("Bundle has no entries, a MessageHeader is required");
+      return oo;
+    }
+
+    // check that there is a message header
+    Resource firstResource = bundleR4.getEntryFirstRep().getResource();
+    if ( firstResource.getResourceType() != ResourceType.MessageHeader) {
+      OperationOutcome oo = new OperationOutcome();
+      oo.addIssue()
+          .setSeverity(OperationOutcome.IssueSeverity.ERROR)
+          .setDiagnostics("First resource in message bundle must be of type 'MessageHeader'");
+      return oo;
+    }
+
+    MessageHeader msgHead = (MessageHeader) firstResource;
+
+    // check messageHeader is well formed
+    if (!msgHead.hasSource()) {
+      OperationOutcome oo = new OperationOutcome();
+      oo.addIssue()
+          .setSeverity(OperationOutcome.IssueSeverity.ERROR)
+          .setDiagnostics("'MessageHeader' is missing source");
+      return oo;
+    } else if (!msgHead.getSource().hasEndpoint()){
+      OperationOutcome oo = new OperationOutcome();
+      oo.addIssue()
+          .setSeverity(OperationOutcome.IssueSeverity.ERROR)
+          .setDiagnostics("'MessageHeader.source' is missing an endpoint");
+      return oo;
+    }
+
+
+    System.out.println(async);
+    if (async.equals("true")) {
```

Comment:
When I omit the async parameter (because it is optional) I get a NullPointerException. I believe it is from this line because async is null